daemon/api.go

```diff
@@ -2637,3 +2639,80 @@ func getLogs(c *Command, r *http.Request, user *auth.UserState) Response {
 		follow:     follow,
 	}
 }
+
+type appInstruction struct {
+	Action string   `json:"action"`
+	Names  []string `json:"names"`
+	client.StartOptions
+	client.StopOptions
+	client.RestartOptions
+}
+
+func postApps(c *Command, r *http.Request, user *auth.UserState) Response {
+	var inst appInstruction
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&inst); err != nil {
+		return BadRequest("cannot decode request body into service operation: %v", err)
+	}
+	if len(inst.Names) == 0 {
+		// on POST, don't allow empty to mean all
+		return BadRequest("cannot perform operation on services without a list of services to operate on")
+	}
+
+	st := c.d.overlord.State()
+	appInfos, rsp := appInfosFor(st, inst.Names, appInfoOptions{service: true})
+	if rsp != nil {
+		return rsp
+	}
+	if len(appInfos) == 0 {
+		// can't happen: appInfosFor with a non-empty list of services
+		// shouldn't ever return an empty appInfos with no error response
+		return InternalError("no services found")
+	}
+
+	argv := make([]string, 2, 3+len(appInfos))
+	argv[0] = "systemctl"
+	argv[1] = inst.Action
+
+	switch inst.Action {
+	case "start":
+		if inst.Enable {
+			argv[1] = "enable"
+			argv = append(argv, "--now")
+		}
+	case "stop":
+		if inst.Disable {
+			argv[1] = "disable"
+			argv = append(argv, "--now")
+		}
+	case "restart":
+		if inst.Reload {
+			argv[1] = "try-reload-or-restart"
+		}
+	default:
+		return BadRequest("unknown action %q", inst.Action)
+	}
+
+	snapNames := make(map[string]bool, len(appInfos))
+	names := make([]string, len(appInfos))
+	for i, svc := range appInfos {
+		argv = append(argv, svc.ServiceName())
+		snapName := svc.Snap.Name()
+		names[i] = snapName + "." + svc.Name
+		snapNames[snapName] = true
+	}
+
+	desc := fmt.Sprintf("%s of %v", inst.Action, names)
+
+	st.Lock()
+	defer st.Unlock()
+	if err := snapstate.CheckChangeConflictMulti(st, snapNames, nil); err != nil {
+		return Conflict(err.Error())
```

Comment:
the target resource is the app (the service). The state's the State :-)

but, to go at this from the other end, what response do you think it should be? It's not a 400, as the request is fine. It's not a 5xx, there's nothing wrong with the server. We could use `422 Unprocessable Entity`,
> The 422 (Unprocessable Entity) status code means the server understands the content type of the request entity (hence a 415(Unsupported Media Type) status code is inappropriate), and the syntax of the request entity is correct (thus a 400 (Bad Request) status code is inappropriate) but was unable to process the contained instructions.